src/EditorFeatures/CSharpTest/TypeInferrer/TypeInferrerTests.cs

```diff
@@ -1080,7 +1086,82 @@ void M()
   }
 }";
 
-            Test(text, "string");
+            Test(text, "System.String");
+        }
+
+        [Fact]
+        [WorkItem(529480)]
+        [Trait(Traits.Feature, Traits.Features.TypeInferenceService)]
+        public void TestCustomCollectionInitializerAddMethod1()
+        {
+            var text =
+@"class C : System.Collections.IEnumerable
+{
+    void M()
+    {
+        var x = new C() { [|a|] };
+    }
+
+    void Add(int i) { }
+    void Add(string s, bool b) { }
+
+    public System.Collections.IEnumerator GetEnumerator()
+    {
+        throw new System.NotImplementedException();
+    }
+}";
+
+            Test(text, "System.Int32", testPosition: false);
+        }
+
+        [Fact]
+        [WorkItem(529480)]
+        [Trait(Traits.Feature, Traits.Features.TypeInferenceService)]
+        public void TestCustomCollectionInitializerAddMethod2()
+        {
+            var text =
+@"class C : System.Collections.IEnumerable
+{
+    void M()
+    {
+        var x = new C() { { ""test"", [|b|] } };
+    }
+
+    void Add(int i) { }
+    void Add(string s, bool b) { }
+
+    public System.Collections.IEnumerator GetEnumerator()
+    {
+        throw new System.NotImplementedException();
+    }
+}";
+
+            Test(text, "System.Boolean");
+        }
+
```

Comment:
Right. On the VB side I didn't try to implement the intellisense scenarios here, just the Generate From Usage scenarios. I'll document that with a new issue.